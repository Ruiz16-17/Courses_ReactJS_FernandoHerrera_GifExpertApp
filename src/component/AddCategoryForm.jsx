import PropTypes from 'prop-types';
import React, { useState } from 'react';

export const AddCategoryForm = ({setCategories}) => {

    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (event) => {
        setInputValue(event.target.value)
    }

    const handleSubmit = (event) => {

        event.preventDefault();

        if(inputValue.trim().length > 2){
            setCategories(categories => [inputValue,...categories]);
            setInputValue('');
        }
        
    }

    return (

        <form onSubmit={handleSubmit}>

            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>

    )
}

AddCategoryForm.propType = {
    setCategories : PropTypes.func.isRequired
}