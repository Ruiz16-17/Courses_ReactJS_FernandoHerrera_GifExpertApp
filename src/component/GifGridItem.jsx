import PropTypes from 'prop-types';
import React from 'react';

export const GifGridItem = ({id, url, title}) => {
    return (
        <div 
            key={id}
            className="card animate__animated animate__bounce"
        >
            <img src={url} alt={title} />
            <p>{title}</p>
        </div>
    )
}

GifGridItem.propType = {
    id : PropTypes.string.isRequired,
    title : PropTypes.string.isRequired,
    url :PropTypes.string.isRequired
    
}