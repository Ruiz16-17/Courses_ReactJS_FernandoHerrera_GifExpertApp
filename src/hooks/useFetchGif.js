import { useEffect, useState } from "react";
import { getGif } from "../helpers/getGifs";

export const useFetchGif = (category) => {
    
    const [state, setState] = useState({
        data: [],
        loading: true
    })

    useEffect(() => {
        getGif(category)
        .then(images => {
            setState({
                data: images,
                loading: false
            })
        });
    }, [category]); //when de void array is present it mean that this will be executed one time

    return state;
}
