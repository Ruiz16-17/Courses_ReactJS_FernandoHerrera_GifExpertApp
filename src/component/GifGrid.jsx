import PropTypes from 'prop-types';
import React from 'react';
import { useFetchGif } from '../hooks/useFetchGif';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    const {loading, data: images} = useFetchGif(category);

    return (
        <>

            <h3 className="animate__animated animate__flash">{category}</h3>

            {loading && <p>Loading...</p> }

            <div className="card-grid">
            
                {images.map(image => {

                    return (
                        <GifGridItem
                            key={image.id}
                            {...image}

                        />
                    )
                })}

            </div>
        </>
    )
}

GifGrid.propType = {
    category : PropTypes.string.isRequired
    
}