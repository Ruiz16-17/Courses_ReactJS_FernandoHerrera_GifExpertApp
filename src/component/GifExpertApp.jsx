import React, { useState } from 'react';
import { AddCategoryForm } from './AddCategoryForm';
import { GifGrid } from './GifGrid';

export const GifExpertApp = ({defaultCategories = []}) => {

    const [categories, setCategories] = useState(defaultCategories);

    return (
        <>
            <h2>GifExpertApp</h2>
            
            <AddCategoryForm setCategories={setCategories}/>
            
            <hr/>

            <ol>
                {
                    categories.map(
                        category => {
                            return <GifGrid 
                                key={category}
                                category={category} />
                        }
                    )
                
                }
            </ol>

        </>
    )
}
