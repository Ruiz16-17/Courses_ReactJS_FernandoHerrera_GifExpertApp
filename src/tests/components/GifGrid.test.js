import '@testing-library/jest-dom';
import { shallow } from "enzyme";
import React from "react";
import { GifGrid } from "../../component/GifGrid";
import { useFetchGif } from "../../hooks/useFetchGif";

jest.mock('../../hooks/useFetchGif');

describe('Test in <GifGrid/>', () => {

    const category = 'One punch man';
    
    test('Should show correctly', () => {
        
        useFetchGif.mockReturnValue({
            data: [],
            loading:true
        });
        
        const wrapper = shallow(<GifGrid category={category}/>);
        expect(wrapper).toMatchSnapshot();
    })
    
    test('should show the items when the images have been loaded useFecthGifs', () => {
        
        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/img.jpg',
            title:'Any thing'
        }]

        useFetchGif.mockReturnValue({
            data: gifs,
            loading:false
        });
        
        const wrapper = shallow(<GifGrid category={category}/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
    })
    
})
