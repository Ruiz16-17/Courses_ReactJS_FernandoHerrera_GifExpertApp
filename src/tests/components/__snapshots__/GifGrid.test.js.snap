// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <GifGrid/> Should show correctly 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__flash"
  >
    One punch man
  </h3>
  <p>
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Test in <GifGrid/> should show the items when the images have been loaded useFecthGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__flash"
  >
    One punch man
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Any thing"
      url="https://localhost/img.jpg"
    />
  </div>
</Fragment>
`;
